stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -760,25 +761,39 @@ extension ${Self} {
 extension ${Self} {
 % for srcBits in allFloatBits:
 %   That = floatName[srcBits]
-%   if Self != That:
 
-%     if srcBits == 80:
+%   if srcBits == 80:
 #if !os(Windows) && (arch(i386) || arch(x86_64))
-%     end
+%   end
 
+%   if srcBits == bits:
+  /// Create an instance initialized to `value`.
+%   else:
   /// Construct an instance that approximates `other`.
+%   end
   public init(_ other: ${That}) {
-%     if srcBits > bits:
+%   if srcBits > bits:
     _value = Builtin.fptrunc_FPIEEE${srcBits}_FPIEEE${bits}(other._value)
-%     else:
+%   elif srcBits < bits:
     _value = Builtin.fpext_FPIEEE${srcBits}_FPIEEE${bits}(other._value)
-%     end
+%   else:
+    _value = other._value
+%   end
+  }
+  
+  /// Create an ${That} initialized to `value`, if `value` can be represented without rounding.
+  @_transparent
+  public init?(exactly other: ${That}) {
+    self.init(other)
+%   if srcBits > bits:
+    if ${That}(self) != other {
+      return nil
+    }
+%   endif
```

Comment:
Yeah I've got that fixed locally. I was waiting on ideas on NaN handling before I pushed changes again.